bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java

```diff
@@ -496,7 +496,7 @@ public Object next() {
 
         @Override
         public boolean hasNext() {
-           return cursor < noOfAddedEntries;
+           return cursor < noOfAddedEntries && values.size() != 0;
         }
```

Comment:
```ballerina
type User record {
    readonly string username;
    string firstName;
};

// @test:Config
function tableRemoveProblem() {
    table<User> key(username) users = table [
        {username: "user1", firstName: "John"},
        {username: "user2", firstName: "Jane"}
    ];

    // Remove the members
    _ = users.remove("user1");
    _ = users.remove("user2");
    foreach var user in users {
    }
}

public function main() {
    tableRemoveProblem();
}
```
IMO this is not the correct fix. when considering the above example,  after removing `user1` and `user2`, `noOfAddedEntries` should be equal to zero.
In here https://github.com/ballerina-platform/ballerina-lang/blob/1538dc17bc415ebe947df159f17cbad79b80830f/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TableValueImpl.java#L675
we do a check before  `noOfAddedEntries` decrement. Do we need this check?
@gimantha WDYT?

Reply:
There are some requirements from iterator implementation coming from the spec. You cannot remove the members which are not already iterated. Any mutations to members since the iterator is created should panic. To track the point where the current cursor is at, the `noOfAddedEntries` is used. Thats not an exact representation of the existing number of members in the table.